File: src/python/xerxes/memory/plugins/mem0.py
# Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Mem0 cloud memory provider (https://mem0.ai)."""

from __future__ import annotations

import os
from typing import Any

from ._base import ExternalMemoryProviderBase


class Mem0Provider(ExternalMemoryProviderBase):
    """Memory provider backed by mem0.ai's HTTP API.

    Pure HTTP (no SDK dependency). Requires ``MEM0_API_KEY`` and
    optionally honours ``MEM0_USER_ID`` (defaults to ``"xerxes"``)."""

    name = "mem0"
    namespace_label = "mem0"
    required_module = None  # Pure HTTP — no SDK required.
    required_env = ("MEM0_API_KEY",)

    BASE_URL = "https://api.mem0.ai/v1"

    def _client(self):
        """Return an authenticated ``httpx.Client`` for the mem0 API."""
        import httpx

        return httpx.Client(
            base_url=self.BASE_URL,
            headers={"Authorization": f"Bearer {os.environ['MEM0_API_KEY']}"},
            timeout=30.0,
        )

    def _call_upstream(self, action: str, arguments: dict[str, Any]) -> Any:
        """Translate a standard action to the matching mem0 HTTP endpoint."""
        user_id = os.environ.get("MEM0_USER_ID", "xerxes")
        with self._client() as c:
            if action == "add":
                resp = c.post(
                    "/memories",
                    json={"messages": [{"role": "user", "content": arguments["content"]}], "user_id": user_id},
                )
                resp.raise_for_status()
                return resp.json()
            if action == "search":
                resp = c.post("/memories/search", json={"query": arguments["query"], "user_id": user_id})
                resp.raise_for_status()
                return resp.json()
            if action == "list":
                resp = c.get("/memories", params={"user_id": user_id, "limit": arguments.get("limit", 20)})
                resp.raise_for_status()
                return resp.json()
            if action == "remove":
                resp = c.delete(f"/memories/{arguments['entry_id']}")
                resp.raise_for_status()
                return {"removed": True}
        raise ValueError(f"unknown mem0 action: {action}")


PROVIDER = Mem0Provider()

__all__ = ["PROVIDER", "Mem0Provider"]

File: xerxes/src/memory/plugins/mem0.ts
// Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
// Licensed under the Apache License, Version 2.0.

import {
  ExternalMemoryProviderBase,
  memoryPluginEnvironmentValue,
  memoryPluginLimit,
  requiredMemoryPluginArgument,
  requiredMemoryPluginEnvironment,
  type MemoryPluginEnvironment,
} from './base.js'
import { HttpMemoryUpstream, memoryPluginPathSegment, memoryPluginUrl, type MemoryPluginHttpTransport } from './http.js'

/** Default Mem0 cloud API endpoint. An embedding application may override it explicitly. */
export const DEFAULT_MEM0_BASE_URL = 'https://api.mem0.ai/v1'

/** Explicit Mem0 configuration and host-owned HTTP transport. */
export interface Mem0ProviderOptions {
  readonly baseUrl?: string
  readonly clock?: () => number
  readonly environment: MemoryPluginEnvironment
  readonly transport: MemoryPluginHttpTransport
}

/** Mem0 HTTP adapter with no global fetch, SDK import, or process-environment lookup. */
export class Mem0Provider extends ExternalMemoryProviderBase {
  readonly baseUrl: string

  constructor(options: Mem0ProviderOptions) {
    const baseUrl = options.baseUrl ?? DEFAULT_MEM0_BASE_URL
    super({
      name: 'mem0',
      namespaceLabel: 'mem0',
      requiredEnvironment: ['MEM0_API_KEY'],
      environment: options.environment,
      upstream: new HttpMemoryUpstream({
        providerName: 'mem0',
        transport: options.transport,
        requestFor: (action, arguments_) => {
          const apiKey = requiredMemoryPluginEnvironment(options.environment, 'MEM0_API_KEY')
          const userId = memoryPluginEnvironmentValue(options.environment, 'MEM0_USER_ID', 'xerxes')
          const headers = { Authorization: `Bearer ${apiKey}` }
          if (action === 'add') {
            return {
              method: 'POST',
              url: memoryPluginUrl(baseUrl, 'memories'),
              headers,
              body: {
                messages: [{ role: 'user', content: requiredMemoryPluginArgument(arguments_, 'content') }],
                user_id: userId,
              },
            }
          }
          if (action === 'search') {
            return {
              method: 'POST',
              url: memoryPluginUrl(baseUrl, 'memories/search'),
              headers,
              body: { query: requiredMemoryPluginArgument(arguments_, 'query'), user_id: userId },
            }
          }
          if (action === 'list') {
            return {
              method: 'GET',
              url: memoryPluginUrl(baseUrl, 'memories', { user_id: userId, limit: memoryPluginLimit(arguments_, 20) }),
              headers,
            }
          }
          const entryId = requiredMemoryPluginArgument(arguments_, 'entry_id')
          return {
            method: 'DELETE',
            url: memoryPluginUrl(baseUrl, `memories/${memoryPluginPathSegment(entryId)}`),
            headers,
          }
        },
      }),
      ...(options.clock === undefined ? {} : { clock: options.clock }),
    })
    this.baseUrl = baseUrl
  }
}
